Implement DatabaseManager::SetDefaultDatabase

Flogex opened the pull request:
The `DatabaseManager::default_database` is returned by `DatabaseManager::GetDefaultDatabase` when the default catalog in the search path is invalid. In this PR, a method `SetDefaultDatabase` is added which can be used to change `default_database`.

With this change, new `Connection`s can use a database as default other than the one that was added first to the `DatabaseManager`.

The value for `default_database` can only be changed internally by calling `SetDefaultDatabase`. There is no `SET GLOBAL` option for this implemented in this PR.

I have written a unit test for this, but did not know where to put it. If you could tell me a good location for it, I can add it to the PR, too.

Flogex committed:
Implement DatabaseManager::SetDefaultDatabase

The `DatabaseManager::default_database` is returned by `DatabaseManager::GetDefaultDatabase`
when the default catalog in the search path is invalid. This commit adds a method to change
`default_database`.
This makes it possible that new `Connection`s can use a database other than 'memory' as default.

Flogex commented on src/main/database_manager.cpp:
```diff
@@ -87,6 +87,29 @@ const string &DatabaseManager::GetDefaultDatabase(ClientContext &context) {
 }
 
+void DatabaseManager::SetDefaultDatabase(ClientContext &context, const string &new_value) {
+	auto has_active_transaction = context.transaction.HasActiveTransaction();
+	if (!has_active_transaction) {
+		context.transaction.BeginTransaction();
+	}
+
+	auto db_entry = GetDatabase(context, new_value);
+
+	if (!has_active_transaction) {
+		context.transaction.ClearTransaction();
+	}
```
I want to validate here that the database referred to by `new_value` actually exists. However, `GetDatabase` requires a transaction to be active. What is the best way to create a transaction when non is active? Or is there another way to validate that the database exists which does not require to start a transaction?

Flogex committed:
Use 'ClientContext::RunFunctionInTransaction' in 'SetDefaultDatabase'

Mytherin commented on src/main/database_manager.cpp:
```diff
@@ -87,6 +87,22 @@ const string &DatabaseManager::GetDefaultDatabase(ClientContext &context) {
+void DatabaseManager::SetDefaultDatabase(ClientContext &context, const string &new_value) {
+	optional_ptr<AttachedDatabase> db_entry;
+
+	context.RunFunctionInTransaction([&]() { db_entry = GetDatabase(context, new_value); });
```
`RunFunctionInTransaction`  is very heavy-handed. It launches a new transaction if none is active (and commits it afterwards). Generally almost all code in DuckDB (be it parser, planner, optimizer, executor) is run with an active transaction already. Is this necessary only for the unittest? I imagine that when this function is used internally it will be used in locations where a transaction already exists.

Flogex replied:
Indeed I added this to make a unit test case succeed in which I did not start a transaction before calling `SetDefaultDatabase`. Otherwise, an `InternalException` with message "TransactionContext::ActiveTransaction called without active transaction" would be thrown. I thought that it would be worth avoiding this exception by starting a new transaction, considering that `RunFunctionInTransaction` does not have a significant overhead when there is already an active transaction, does it?

Do you want me to remove `RunFunctionInTransaction` and just call `GetDatabase` directly?

Mytherin replied:
Yes, most functions in the catalog rely on a transaction existing. I think it is better to remove it.

Mytherin reviewed (commented):
Thanks for the PR! Looks good to me - one comment:

Flogex committed:
Removed call to 'RunFunctionInTransaction' in 'DatabaseManager::SetDefaultDatabase'

Most functions that call 'SetDefaultDatabase' will already have an active transaction,
hence it is not necessary to create a new one.

Mytherin committed:
Update database_manager.cpp
